src/Compilers/CSharp/Portable/FlowAnalysis/ControlFlowPass.cs

```diff
@@ -400,19 +453,65 @@ public override BoundNode VisitBlock(BoundBlock node)
             var parentBlock = _currentBlock;
             _currentBlock = node;
             var initialUsingCount = _usingDeclarations.Count;
+            var hadUsingDeclarations = false;
             foreach (var local in node.Locals)
             {
                 if (local.IsUsing)
                 {
                     _usingDeclarations.Add((local, node));
+                    hadUsingDeclarations = true;
                 }
             }
 
             var result = base.VisitBlock(node);
 
             _usingDeclarations.Clip(initialUsingCount);
+            if (_asyncTryFinallyEndsReachable is not null && hadUsingDeclarations)
+            {
+                UsingDeclarationVisitor.Visit(node, _asyncTryFinallyEndsReachable, State.Alive);
+            }
             _currentBlock = parentBlock;
             return result;
         }
+
+        private sealed class UsingDeclarationVisitor : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
```

Comment:
From offline discussion, consider cutting off when running into an expression. Then we can use a more basic basic type